### PR TITLE
[python] [orttraining] Add utility to export a graph to compute gradients

### PR DESCRIPTION
**Pull request by juharris**
**Description**:
Add a Python utility to create a gradient graph that can help with training in other environments (like JavaScript).
Input: model, loss function, example input, example labels, graph output path, some config
Result: A gradient graph is written to the output path for a model that takes in: model input, target label, model parameters/weights and outputs gradients for the model parameters

**Motivation and Context**
- The gradient graph can be loaded in various environments (like JavaScript) and facilitate training models in a training loop.

TODO:
- [x] Add more documentation.
- [x] I made this to see if the general idea and direction is okay. I'll wrap some of the Python logic in a helper function to wrap the `Module` in one that uses the `label` and `loss_fn` and also hides the `register_all` call.
- [x] Add Python tests (where should I add them? I saw some tests in orttraining/orttraining/test/python, is there a file that I should copy as an example?)
- [x] Clean up pointer/reference usage? Sorry, it's been a while since I've done C++ so please let me know if I need to correct some pointer/reference usage, I'm happy to learn.
- [x] Override constructors/initialize more? I just allowed one way to set up the Python binding but once I get this working, we can do others, such as taking in a serialized model string.

### CODE REVIEW

**Comment by juharris** [comment]
I'm getting errors when trying to build:
```
./build.sh --config RelWithDebInfo --build_shared_lib --parallel --enable_training --enable_pybind  --skip_submodule_sync
```
...
```
2021-06-24 16:33:53,250 util.run [INFO] - Running subprocess in '/home/juharri/workspace/onnx/onnxruntime/build/Linux/RelWithDebInfo'
['/home/juharri/miniconda3/envs/ort-dev/bin/python3', 'orttraining_test_gradient_graph.py']
2021-06-24 15:33:53.737855800 [E:onnxruntime:Default, provider_bridge_ort.cc:947 Ensure] Failed to load library libonnxruntime_providers_shared.so with error: libonnxruntime_providers_shared.so: cannot open shared object file: No such file or directory
2021-06-24 15:33:53.737925600 [W:onnxruntime:Default, onnxruntime_pybind_state.cc:1678 CreatePybindStateModule] Init provider bridge failed.
2021-06-24 16:33:54,629 util.run [DEBUG] - Subprocess completed. Return code: 0
2021-06-24 16:33:54,630 util.run [INFO] - Running subprocess in '/home/juharri/workspace/onnx/onnxruntime/build/Linux/RelWithDebInfo'
['/home/juharri/miniconda3/envs/ort-dev/bin/python3', 'onnxruntime_test_python_mlops.py']
2021-06-24 15:33:54.874918400 [E:onnxruntime:Default, provider_bridge_ort.cc:947 Ensure] Failed to load library libonnxruntime_providers_shared.so with error: libonnxruntime_providers_shared.so: cannot open shared object file: No such file or directory
2021-06-24 15:33:54.874985000 [W:onnxruntime:Default, onnxruntime_pybind_state.cc:1678 CreatePybindStateModule] Init provider bridge failed.
```

Yet the exit code is still 0. What should I do to get my test (orttraining_test_gradient_graph.py) to run?

**Comment by ryotatomioka** [comment]
I had a similar thought but I was wondering if we could use `OrtModuleGraphBuilder` for this purpose.

**Comment by juharris** [comment]
@ryotatomioka That would be nice. I asked @SherlockNoMad about that while working on this and he said:
> the logic in ortmodule_graph_builder.cc is specific to ORTModule, and the graph produced won't be runable with InferenceSession

File: orttraining/orttraining/core/framework/gradient_graph_builder.cc [hunk 1]
```diff
@@ -89,6 +89,7 @@ GradientGraphBuilder::GradientGraphBuilder(Graph* graph,
   std::string unreachable_nodes;
   // building x_nodes_
   for (const auto& name : x_node_arg_names) {
+    std::cout << "Adding x_node_arg_names: name: " << name << std::endl;
```
Comment (juharris) [inline orttraining/orttraining/core/framework/gradient_graph_builder.cc:92]: I'll remove these print statements before merging.

**Comment by xadupre** [comment]
What is the different with [get_model()](https://github.com/microsoft/onnxruntime/blob/master/orttraining/orttraining/python/orttraining_pybind_state.cc#L751) and [get_inference_optimized_model](https://github.com/microsoft/onnxruntime/blob/master/orttraining/orttraining/python/orttraining_pybind_state.cc#L755)?

**Comment by juharris** [comment]
Thanks for looking into this @xadupre. Sorry, I'll update the title of this PR to clarify how the direction of this has changed slightly. I think I did try to do something with those first or some other alternatives but those just graphs just output gradients and I think they don't take the model weights as inputs, they use the model weights that must already be in the graph.

I want to facilitate training in other languages (like in JavaScript) so what's useful is a graph that you can use in a training loop that will compute updated weights given the previous weights. BTW I managed to get the model parameters as inputs and I'd like to configure an optimizer and have the graph output the updated model weights. Do you know how I can do that? Maybe @SherlockNoMad and @cianmhayes can help clarify further.

**Comment by xadupre** [comment]
Hi @juharris, I did something in python if you are interesting. My goal was to use orttraining to implement stochastic gradient descent the same way scikit-learn is doing, using the same mechanism implemented in ORTModule but without pytorch. I wrote a documentation about this: [Partial Training with OrtGradientForwardBackwardOptimizer](http://www.xavierdupre.fr/app/onnxcustom/helpsphinx/tutorial_training/tutorial_6_training_partial.html) and [Training with onnxruntime](http://www.xavierdupre.fr/app/onnxcustom/helpsphinx/tutorial_onnxruntime/training_ort_api.html). It works for a regression.

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 2]
```diff
@@ -760,6 +772,40 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
         return ortmodule_graph_builder->GetGraphInfo();
       });
 
+  py::class_<PyGradientGraphBuilder> gradient_graph_builder(m, "GradientGraphBuilder", R"pbdoc(A utility for making a gradient graph that can be used to help train a model.)pbdoc");
```
Comment (baijumeswani) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:775]: Would you please add a code comment here explaining the need for this pybind API?

File: orttraining/orttraining/python/training/gradient_graph/_gradient_graph_tools.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
+from pathlib import Path
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/gradient_graph/_gradient_graph_tools.py:1]: May we move this file to orttraining/orttraining/python/training/**experimental**/gradient_graph/_gradient_graph_tools.py?

File: orttraining/orttraining/python/training/__init__.py [hunk 4]
```diff
@@ -3,13 +3,14 @@
 # Licensed under the MIT License.
 # --------------------------------------------------------------------------
 
-from onnxruntime.capi._pybind_state import TrainingParameters
-from onnxruntime.capi._pybind_state import PropagateCastOpsStrategy
+from onnxruntime.capi._pybind_state import GradientGraphBuilder, PropagateCastOpsStrategy, TrainingParameters
 from onnxruntime.capi.training.training_session import TrainingSession
 
+from .gradient_graph._gradient_graph_tools import export_gradient_graph
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/__init__.py:9]: Let's not add this here. Let uses explicitly call into the experimental module to call `export_gradient_graph`.

File: tools/ci_build/build.py [hunk 5]
```diff
@@ -1598,6 +1598,7 @@ def run_onnxruntime_tests(args, source_dir, ctest_path, build_dir, configs):
                 cwd = os.path.join(cwd, config)
 
             run_subprocess([sys.executable, 'onnxruntime_test_python.py'], cwd=cwd, dll_path=dll_path)
+            run_subprocess([sys.executable, 'orttraining_test_gradient_graph.py'], cwd=cwd, dll_path=dll_path)
```
Comment (baijumeswani) [inline tools/ci_build/build.py:1601]: Let's add a comment here explaining what this test is.
Let's also move it under the method: `run_training_python_frontend_tests`

File: orttraining/orttraining/test/python/orttraining_test_gradient_graph.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
+import os
```
Comment (baijumeswani) [inline orttraining/orttraining/test/python/orttraining_test_gradient_graph.py:1]: Let's add experimental in the test file name: `orttraining_test_experimental_gradient_graph.py`

File: orttraining/orttraining/python/orttraining_pybind_state.cc (hunk 2 above)
Comment (juharris) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:777]: You mean comments for the C++ file right? I added some.

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 7]
```diff
@@ -760,6 +772,42 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
...
+            loss_node_arg_name,
+            *gradient_graph_config_ptr,
+            *logger_ptr);
+
+        return std::make_unique<PyGradientGraphBuilder>(std::move(builder), std::move(model), std::move(logger_ptr), std::move(gradient_graph_config_ptr));
+      }))
+      .def("build", [](PyGradientGraphBuilder* gradient_graph_builder) {
+        ORT_THROW_IF_ERROR(gradient_graph_builder->builder->Build());
+      })
+      .def("save", [](PyGradientGraphBuilder* gradient_graph_builder, const std::string& path) {
```
Comment (xadupre) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:807]: Same comment as above, it is possible to get the onnx graph instead of saving it?

File: orttraining/orttraining/python/training/experimental/gradient_graph/_gradient_graph_tools.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+    if not isinstance(intermediate_graph_path, str):
+        intermediate_graph_path = str(intermediate_graph_path)
+
+    if not isinstance(gradient_graph_path, str):
+        gradient_graph_path = str(gradient_graph_path)
+
+    class WrapperModule(torch.nn.Module):
+        def forward(self, model_input, expected_labels, *model_params):
+                for param, set_param in zip(model.parameters(), model_params):
```
Comment (xadupre) [inline orttraining/orttraining/python/training/experimental/gradient_graph/_gradient_graph_tools.py:56]: indentation?

File: orttraining/orttraining/python/training/experimental/gradient_graph/_gradient_graph_tools.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+        gradient_graph_path (Union[Path, str]): The path to where you would like to save the gradient graph.
+
+        intermediate_graph_path (Optional[Union[Path, str]): The path to where you would like to save any intermediate graphs that are needed to make the gradient graph.
+        Defaults to `gradient_graph_path` and it will be overwritten if this function executes successfully.
+
+        opset_version (int): See `torch.onnx.export`.
+    """
+
+    # Is this needed?
```
Comment (xadupre) [inline orttraining/orttraining/python/training/experimental/gradient_graph/_gradient_graph_tools.py:39]: Is this needed?

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 10]
```diff
@@ -760,6 +772,42 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
         return ortmodule_graph_builder->GetGraphInfo();
       });
 
+  // Provide a convenient and well-documented way to make a gradient graph.
+  // It's possible to get the gradient graph through ORTModule by leveraging some "private" fields and not-so-well-documented APIs, so we provide this explicit and tested way to get the gradient graph.
+  py::class_<PyGradientGraphBuilder> gradient_graph_builder(m, "GradientGraphBuilder", R"pbdoc(A utility for making a gradient graph that can be used to help train a model.)pbdoc");
+  // Set up methods to match the C++ `GradientGraphBuilder` interface.
+  gradient_graph_builder.def(py::init([](
```
[the remark this answers is not in the document]
Comment (juharris) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:779]: Sorry I'm not actually that familiar with this code base, is there a specific type that you had in mind for "onnx graph"? That would be great if the developer didn't have to `torch.onnx.export(...` to a file.

As for the idea itself, I'm not sure how useful that will be since this is to expose a convenient way to use `GradientGraphBuilder` to Python. `GradientGraphBuilder` does take in `onnxruntime::Graph`.

File: orttraining/orttraining/python/orttraining_pybind_state.cc (hunk 7 above)
Comment (juharris) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:807]: That would be cool. I built this to match the `GradientGraphBuilder` in C++ which has this `build(..)`, `save(...)` interface but I also thought that it's good because once `build` is called, the graph can be manipulated with methods that don't exist yet. Yeah it sucks to force developers to use a file if it's not really needed but if you're calling this method, then you probably do want the file, eventually.

File: orttraining/orttraining/python/training/experimental/gradient_graph/_gradient_graph_tools.py (hunk 9 above)
Comment (juharris) [reply orttraining/orttraining/python/training/experimental/gradient_graph/_gradient_graph_tools.py]: Good catch. It's not needed and I removed it. Sorry, there's still some things I'm checking.

File: orttraining/orttraining/python/orttraining_pybind_state.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (xadupre) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:779]: It is sometime useful to avoid writing onnx file on disk and it is faster for unit tests.

File: orttraining/orttraining/python/training/experimental/gradient_graph/_gradient_graph_tools.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+        'labels': {0: 'batch_size', },
+        'output': {0: 'batch_size', },
+    }
+
+    args = (example_input, example_labels, *tuple(model.parameters()))
+    model_param_names = tuple(name for name, _ in model.named_parameters())
+    input_names = ['input', 'labels', *model_param_names]
+    nodes_needing_gradients = set()
+    for name, _ in model.named_parameters():
+        # Should we check `if _.requires_grad:` first?
```
Comment (juharris) [inline orttraining/orttraining/python/training/experimental/gradient_graph/_gradient_graph_tools.py:72]: Yes we should add this check.

File: orttraining/orttraining/python/orttraining_pybind_state.cc (hunk 7 above)
Comment (juharris) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:809]: I think it's good to support exporting to a file for now. We can add support in the future for exporting the model to other formats, such as a string so that it can be loaded by `onnx.load_model_from_string`.

File: orttraining/orttraining/test/python/orttraining_test_experimental_gradient_graph.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+
+def binary_cross_entropy_loss(inp, target):
+    loss = -torch.sum(target * torch.log2(inp[:, 0]) +
+                      (1-target) * torch.log2(inp[:, 1]))
+    return loss
+
+
+class GradientGraphBuilderTest(unittest.TestCase):
+    def test_save(self):
```
Comment (baijumeswani) [inline orttraining/orttraining/test/python/orttraining_test_experimental_gradient_graph.py:52]: We could probably change the name of this test. I think let's address this separately later on.

**Review by baijumeswani — APPROVED** [review]
Looks good.

File: orttraining/orttraining/test/python/orttraining_test_experimental_gradient_graph.py (hunk 12 above)
Comment (juharris) [reply orttraining/orttraining/test/python/orttraining_test_experimental_gradient_graph.py:52]: I don't might changing it but it's testing the `save` method of `GradientGraphBuilder` so it makes sense when you look at the test class name.

Comment (baijumeswani) [reply orttraining/orttraining/test/python/orttraining_test_experimental_gradient_graph.py:52]: I feel the test is doing more than just saving (it actually executes the `InferenceSession`). I think it is ok to leave it as `save` for now.
